### PR TITLE
Enable pyTorch-IMage-Models (TIMM) with HPUs

### PR DESCRIPTION
**Pull request by ZhengHongming888**
# What does this PR do?

This PR contains the scripts that showcases how to inference/fine-tune the TIMM models on intel's HPUs with the lazy/graph modes.  We support the training for single/multiple HPU cards both two. Currently we support several most downloadable models from Hugging Face as below list.

- [timm/resnet50.a1_in1k](https://huggingface.co/timm/resnet50.a1_in1k)
- [timm/resnet18.a1_in1k](https://huggingface.co/timm/resnet18.a1_in1k)
- [timm/resnet18.fb_swsl_ig1b_ft_in1k](https://huggingface.co/timm/resnet18.fb_swsl_ig1b_ft_in1k)
...
- [timm/vit_base_patch16_224.augreg2_in21k_ft_in1k](https://huggingface.co/timm/vit_base_patch16_224.augreg2_in21k_ft_in1k)
- [timm/vgg19.tv_in1k]()

Here we support the below features: 

1. Single-HPU training
Here we show how to fine-tune the [imagenette2-320 dataset](https://www.kaggle.com/datasets/xbinchen/imagenette2-320) and model with [timm/resnet50.a1_in1k](https://huggingface.co/timm/resnet50.a1_in1k) from Hugging Face.

1) training with hpu lazy mode
   
```bash
python train_hpu_lazy.py \
    --data-dir ./imagenette2-320/ \
    --device 'hpu' \
    --model resnet50.a1_in1k
```
2) training with hpu graph mode

```bash
python train_hpu_graph.py \
    --data-dir ./imagenette2-320/ \
    --device 'hpu' \
    --model resnet50.a1_in1k
```

2. Multi-HPU training

1) training with hpu lazy mode
```bash
torchrun --nnodes 1 --nproc_per_node 2 \
    train_hpu_lazy.py \
    --data-dir ./imagenette2-320/ \
    --device 'hpu' \
    --model resnet50.a1_in1k
```

2) training with hpu graph mode

```bash
torchrun --nnodes 1 --nproc_per_node 2 \
    train_hpu_graph.py \
    --data-dir ./imagenette2-320/ \
    --device 'hpu' \
    --model resnet50.a1_in1k
```

3. Single HPU inference

1)  hpu with graph_mode
```bash
python inference.py \
    --data-dir='./download_ds/imagenette2-320' \
    --device='hpu' \
    --model resnet50.a1_in1k \
    --graph_mode
```

2) hpu with lazy mode
```bash
python inference.py \
    --data-dir='./download_ds/imagenette2-320' 
    --device='hpu' \
    --model resnet50.a1_in1k
```
Welcome for  any suggestions/comments. Thanks.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by ssarkar2 — APPROVED** [review]

File: examples/pytorch-image-models/README.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+pip install .
+```
+
+## Single-HPU training
+
+### Using datasets from Hub
+
+Here we show how to fine-tune the [imagenette2-320 dataset](https://www.kaggle.com/datasets/xbinchen/imagenette2-320) and model with [timm/resnet50.a1_in1k](https://huggingface.co/timm/resnet50.a1_in1k) from Hugging Face.
+
+### training with hpu lazy mode
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:49]:
```suggestion
### Training with HPU lazy mode
```

File: examples/pytorch-image-models/README.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+Here the results for lazy mode is shown below for example:
+
+
+Train: 0 [   0/73 (  1%)]  Loss: 6.86 (6.86)  Time: 9.575s,   13.37/s  (9.575s,   13.37/s)  LR: 1.000e-05  Data: 0.844 (0.844)
+
+Train: 0 [  50/73 ( 70%)]  Loss: 6.77 (6.83)  Time: 0.320s,  400.32/s  (0.470s,  272.39/s)  LR: 1.000e-05  Data: 0.217 (0.047)
+
+Test: [   0/30]  Time: 6.593 (6.593)  Loss:   6.723 ( 6.723)  Acc@1:   0.000 (  0.000)  Acc@5:   0.000 (  0.000)
+
+Test: [  30/30]  Time: 3.856 (0.732)  Loss:   6.615 ( 6.691)  Acc@1:   0.000 (  0.076)  Acc@5:   1.176 (  3.287)
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:75]: It's not very readable this way, maybe put this in code block with ``` ?

File: examples/pytorch-image-models/README.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+Train: 1 [  50/73 ( 70%)]  Loss: 3.23 (3.76)  Time: 0.160s,  798.85/s  (0.148s,  863.22/s)  LR: 1.001e-02  Data: 0.053 (0.051)
+
+Test: [   0/30]  Time: 0.663 (0.663)  Loss:   1.926 ( 1.926)  Acc@1:  46.094 ( 46.094)  Acc@5:  85.938 ( 85.938)
+
+Test: [  30/30]  Time: 0.022 (0.126)  Loss:   1.462 ( 1.867)  Acc@1:  63.529 ( 39.261)  Acc@5:  83.529 ( 85.096)
+
+Current checkpoints:
+ ('./output/train/20241016-034927-resnet50_a1_in1k-224/checkpoint-1.pth.tar', 39.26114640448503)
+ ('./output/train/20241016-034927-resnet50_a1_in1k-224/checkpoint-0.pth.tar', 0.07643312101910828)
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:90]: Same question

File: examples/pytorch-image-models/README.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+
+
+## Multi-HPU training
+
+Here we show how to fine-tune the [imagenette2-320 dataset](https://www.kaggle.com/datasets/xbinchen/imagenette2-320) and model with [timm/resnet50.a1_in1k](https://huggingface.co/timm/resnet50.a1_in1k) from Hugging Face.
+
+### training with hpu lazy mode
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:98]:
```suggestion
### Training with HPU lazy mode
```

File: examples/pytorch-image-models/README.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+```bash
+torchrun --nnodes 1 --nproc_per_node 2 \
+    train_hpu_lazy.py \
+    --data-dir ./imagenette2-320/ \
+    --device 'hpu' \
+    --model resnet50.a1_in1k
+```
+### training with hpu graph mode
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:106]:
```suggestion
### Training with HPU graph mode
```

File: examples/pytorch-image-models/README.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+Here the results for lazy mode is shown below for example:
+
+Train: 0 [   0/36 (  3%)]  Loss: 6.88 (6.88)  Time: 10.036s,   25.51/s  (10.036s,   25.51/s)  LR: 1.000e-05  Data: 0.762 (0.762)
+
+Distributing BatchNorm running means and vars
+
+Test: [   0/15]  Time: 7.796 (7.796)  Loss:   6.915 ( 6.915)  Acc@1:   0.000 (  0.000)  Acc@5:   0.000 (  0.000)
+
+Test: [  15/15]  Time: 1.915 (1.263)  Loss:   6.847 ( 6.818)  Acc@1:   0.000 (  0.000)  Acc@5:   0.000 (  0.688)
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:124]: Same

File: examples/pytorch-image-models/README.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+Current checkpoints:
+ ('./output/train/20241016-034443-resnet50_a1_in1k-224/checkpoint-0.pth.tar', 0.0)
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:127]: Same

File: examples/pytorch-image-models/README.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+Train: 1 [   0/36 (  3%)]  Loss: 6.84 (6.84)  Time: 6.687s,   38.28/s  (6.687s,   38.28/s)  LR: 2.001e-02  Data: 0.701 (0.701)
+
+Distributing BatchNorm running means and vars
+
+Test: [   0/15]  Time: 1.315 (1.315)  Loss:   2.463 ( 2.463)  Acc@1:  14.062 ( 14.062)  Acc@5:  48.828 ( 48.828)
+
+Test: [  15/15]  Time: 0.020 (0.180)  Loss:   1.812 ( 1.982)  Acc@1:  52.326 ( 32.934)  Acc@5:  66.279 ( 75.064)
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:135]: Same

File: examples/pytorch-image-models/README.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+Current checkpoints:
+ ('./output/train/20241016-034443-resnet50_a1_in1k-224/checkpoint-1.pth.tar', 32.93428432485976)
+ ('./output/train/20241016-034443-resnet50_a1_in1k-224/checkpoint-0.pth.tar', 0.0)
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:139]: Same

File: examples/pytorch-image-models/README.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+
+
+## Single-HPU inference
+
+Here we show how to fine-tune the [imagenette2-320 dataset](https://www.kaggle.com/datasets/xbinchen/imagenette2-320) and model with [timm/resnet50.a1_in1k](https://huggingface.co/timm/resnet50.a1_in1k) from Hugging Face.
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:145]:
```suggestion
Here we show how to fine-tune the [imagenette2-320 dataset](https://huggingface.co/datasets/johnowhitaker/imagenette2-320) and model with [timm/resnet50.a1_in1k](https://huggingface.co/timm/resnet50.a1_in1k) from Hugging Face.
```

File: examples/pytorch-image-models/README.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+
+### hpu with graph_mode
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:147]:
```suggestion
### HPU with graph mode
```

File: examples/pytorch-image-models/README.md [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+```bash
+python inference.py \
+    --data-dir='./download_ds/imagenette2-320' \
+    --device='hpu' \
+    --model resnet50.a1_in1k \
+    --graph_mode
+```
+
+### hpu with lazy mode
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:156]:
```suggestion
### HPU with lazy mode
```

File: examples/pytorch-image-models/README.md [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+```bash
+python inference.py \
+    --data-dir='./download_ds/imagenette2-320' 
+    --device='hpu' \
+    --model resnet50.a1_in1k
+```
+
+Models that have been validated same as training lists supported as above.
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:164]: I would remove this sentence as the list at the beginning appears before training examples

File: examples/pytorch-image-models/README.md [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+
+### Using datasets from Hub
+
+Here we show how to fine-tune the [imagenette2-320 dataset](https://www.kaggle.com/datasets/xbinchen/imagenette2-320) and model with [timm/resnet50.a1_in1k](https://huggingface.co/timm/resnet50.a1_in1k) from Hugging Face.
+
+### training with hpu lazy mode
+   
+```bash
+python train_hpu_lazy.py \
+    --data-dir ./imagenette2-320/ \
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:53]: With the dataset from the hub, we probably don't need a data dir

File: examples/pytorch-image-models/README.md [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+    --device 'hpu' \
+    --model resnet50.a1_in1k
+```
+### training with hpu graph mode
+
+```bash
+python train_hpu_graph.py \
+    --data-dir ./imagenette2-320/ \
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:61]: Same

File: examples/pytorch-image-models/README.md [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+## Multi-HPU training
+
+Here we show how to fine-tune the [imagenette2-320 dataset](https://www.kaggle.com/datasets/xbinchen/imagenette2-320) and model with [timm/resnet50.a1_in1k](https://huggingface.co/timm/resnet50.a1_in1k) from Hugging Face.
+
+### training with hpu lazy mode
+```bash
+torchrun --nnodes 1 --nproc_per_node 2 \
+    train_hpu_lazy.py \
+    --data-dir ./imagenette2-320/ \
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:102]: Same

File: examples/pytorch-image-models/README.md [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+    --device 'hpu' \
+    --model resnet50.a1_in1k
+```
+### training with hpu graph mode
+
+```bash
+torchrun --nnodes 1 --nproc_per_node 2 \
+    train_hpu_graph.py \
+    --data-dir ./imagenette2-320/ \
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:111]: Same

File: examples/pytorch-image-models/README.md [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+
+
+## Single-HPU inference
+
+Here we show how to fine-tune the [imagenette2-320 dataset](https://www.kaggle.com/datasets/xbinchen/imagenette2-320) and model with [timm/resnet50.a1_in1k](https://huggingface.co/timm/resnet50.a1_in1k) from Hugging Face.
+
+### hpu with graph_mode
+```bash
+python inference.py \
+    --data-dir='./download_ds/imagenette2-320' \
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:150]: Same

File: examples/pytorch-image-models/README.md [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+    --device='hpu' \
+    --model resnet50.a1_in1k \
+    --graph_mode
+```
+
+### hpu with lazy mode
+```bash
+python inference.py \
+    --data-dir='./download_ds/imagenette2-320'
```
Comment (regisss) [inline examples/pytorch-image-models/README.md:159]: Same

File: examples/pytorch-image-models/README.md [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+- [timm/resnet18.fb_swsl_ig1b_ft_in1k](https://huggingface.co/timm/resnet18.fb_swsl_ig1b_ft_in1k)
+- [timm/wide_resnet50_2.racm_in1k](https://huggingface.co/timm/wide_resnet50_2.racm_in1k)
+- [timm/efficientnet_b3.ra2_in1k](https://huggingface.co/timm/efficientnet_b3.ra2_in1k)
+- [timm/efficientnet_lite0.ra_in1k](https://huggingface.co/timm/efficientnet_lite0.ra_in1k)
+- [timm/efficientnet_b0.ra_in1k](https://huggingface.co/timm/efficientnet_b0.ra_in1k)
+- [timm/nf_regnet_b1.ra2_in1k](https://huggingface.co/timm/nf_regnet_b1.ra2_in1k)
+- [timm/mobilenetv3_large_100.ra_in1k](https://huggingface.co/timm/mobilenetv3_large_100.ra_in1k)
+- [timm/tf_mobilenetv3_large_minimal_100.in1k](https://huggingface.co/timm/tf_mobilenetv3_large_minimal_100.in1k)
+- [timm/vit_base_patch16_224.augreg2_in21k_ft_in1k](https://huggingface.co/timm/vit_base_patch16_224.augreg2_in21k_ft_in1k)
+- [timm/vgg19.tv_in1k]()
```
[the remark this answers is not in the document]
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: added. thanks

File: examples/pytorch-image-models/README.md (hunk 2 above)
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks.

File: examples/pytorch-image-models/README.md [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+Current checkpoints:
+ ('./output/train/20241016-034927-resnet50_a1_in1k-224/checkpoint-0.pth.tar', 0.07643312101910828)
+
+Train: 1 [   0/73 (  1%)]  Loss: 6.69 (6.69)  Time: 0.796s,  160.74/s  (0.796s,  160.74/s)  LR: 1.001e-02  Data: 0.685 (0.685)
+
+Train: 1 [  50/73 ( 70%)]  Loss: 3.23 (3.76)  Time: 0.160s,  798.85/s  (0.148s,  863.22/s)  LR: 1.001e-02  Data: 0.053 (0.051)
+
+Test: [   0/30]  Time: 0.663 (0.663)  Loss:   1.926 ( 1.926)  Acc@1:  46.094 ( 46.094)  Acc@5:  85.938 ( 85.938)
+
+Test: [  30/30]  Time: 0.022 (0.126)  Loss:   1.462 ( 1.867)  Acc@1:  63.529 ( 39.261)  Acc@5:  83.529 ( 85.096)
```
[the remark this answers is not in the document]
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks.

File: examples/pytorch-image-models/README.md [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+Train: 0 [   0/73 (  1%)]  Loss: 6.86 (6.86)  Time: 9.575s,   13.37/s  (9.575s,   13.37/s)  LR: 1.000e-05  Data: 0.844 (0.844)
+
+Train: 0 [  50/73 ( 70%)]  Loss: 6.77 (6.83)  Time: 0.320s,  400.32/s  (0.470s,  272.39/s)  LR: 1.000e-05  Data: 0.217 (0.047)
+
+Test: [   0/30]  Time: 6.593 (6.593)  Loss:   6.723 ( 6.723)  Acc@1:   0.000 (  0.000)  Acc@5:   0.000 (  0.000)
+
+Test: [  30/30]  Time: 3.856 (0.732)  Loss:   6.615 ( 6.691)  Acc@1:   0.000 (  0.076)  Acc@5:   1.176 (  3.287)
+
+Current checkpoints:
+ ('./output/train/20241016-034927-resnet50_a1_in1k-224/checkpoint-0.pth.tar', 0.07643312101910828)
```
[the remark this answers is not in the document]
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks.

File: examples/pytorch-image-models/README.md (hunk 3 above)
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks.

File: examples/pytorch-image-models/README.md (hunk 6 above)
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks.

File: examples/pytorch-image-models/README.md (hunk 8 above)
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks.

File: examples/pytorch-image-models/README.md (hunk 13 above)
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: removed. thanks

File: examples/pytorch-image-models/README.md (hunk 14 above)
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks

File: examples/pytorch-image-models/README.md (hunk 15 above)
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks

File: examples/pytorch-image-models/README.md (hunk 16 above)
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks

File: examples/pytorch-image-models/README.md (hunk 17 above)
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks

File: examples/pytorch-image-models/README.md (hunk 18 above)
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks

File: examples/pytorch-image-models/README.md (hunk 19 above)
Comment (ZhengHongming888) [reply examples/pytorch-image-models/README.md]: corrected. thanks

**Comment by regisss** [comment]
@ZhengHongming888 Can you add a link to the scripts these examples are inspired of please? You can put it in the docstring at the beginning of each file

**Comment by ZhengHongming888** [comment]
> @ZhengHongming888 Can you add a link to the scripts these examples are inspired of please? You can put it in the docstring at the beginning of each file

@regisss  updated please check! Thanks.

**Review by regisss — COMMENTED** [review]
Finally can you create a `test_timm.py` file in https://github.com/huggingface/optimum-habana/tree/main/tests where you quickly test the example scripts please?
For example just running a few training/inference steps to make sure the examples work.

**Comment by ZhengHongming888** [comment]
> Finally can you create a `test_timm.py` file in https://github.com/huggingface/optimum-habana/tree/main/tests where you quickly test the example scripts please? For example just running a few training/inference steps to make sure the examples work.

Thanks @regisss  Done please check! Thanks.

**Pull request merged**